### PR TITLE
Initialize source code for doip client

### PR DESCRIPTION
**Pull request by huavanthong**
Hi Thomas,

I would like to initialize the code for DoIP Client. 
Please help me check it, if you have any comments, please tell me.

Thank you,
Thong Hua

### CODE REVIEW

**Review by test-fullautomation — CHANGES_REQUESTED** [review]
Hi Thong,
Thank you for your pull-request.

I'm confused now. I would expect that we install the doip client directly from pip as part of requirements.txt.

Here in this repository I would expect the robot keyword wrapper for the doip client library?

If we need to change the doip client sources, then we would need a fork. But i think and hope that this is not required.

Thank you,
 Thomas

**Comment by huavanthong** [comment]
Hi Thomas,

I have created the draft version for the robot keyword wrapper for the doip client library on ticket https://github.com/test-fullautomation/robotframework-doip/issues/3.
Beside that, I also created a lot of test cases using keyword from wrapper. I can execute these test cases successfully when I start up a DoIP simulation.
Please help me review it, if you have any concerns, please let me know.

Thank you,
Thong Hua

File: RobotFramework_DoIP/DoipKeywords.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        logical_address = announcement.logical_address
+        ip, port = address
+        print(ip, port, logical_address)
+
+        self.ip = ip
+        self.logical_address = logical_address
+        logger.info(f"IP: {ip} ")
+
+    @keyword
+    def connect(self):
```
Comment (test-fullautomation) [inline RobotFramework_DoIP/DoipKeywords.py]: Hi Thong,
as per https://python-doipclient.readthedocs.io/en/latest/
DoIPClient has the interface:
(ecu_ip_address, ecu_logical_address, tcp_port=13400, udp_port=13400, activation_type=ActivationType.Default, protocol_version=2, client_logical_address=3584, client_ip_address=None, use_secure=False, auto_reconnect_tcp=False)

This interface need to be available as keyword if the user don't want to use get_entity, too for new connection.

Thank you,
 Thomas

File: RobotFramework_DoIP/DoipKeywords.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    def await_vehicle_announcement(self):
+        address, announcement = DoIPClient.await_vehicle_announcement()
+        print(address)
+        logical_address = announcement.logical_address
+        ip, port = address
+        print(ip, port, logical_address)
+        return ip, port, logical_address
+
+    @keyword("Get Entity")
+    def get_entity(self):
```
Comment (test-fullautomation) [inline RobotFramework_DoIP/DoipKeywords.py]: Hi Thong,
here I miss parameters e.g. 
ecu_ip_address='255.255.255.255', protocol_version=2, eid=None, vin=None

You need to wrap the whole interface inclusive parameters, that the full functionality is available as robot keywords.

Thank you,
 Thomas

File: RobotFramework_DoIP/DoipKeywords.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    @keyword
+    def send_diagnostic(self):
+        pass
+    
+    @keyword
+    def receive_diagnostic(self):
+        pass
+
+    @keyword
+    def reconnect(self):
```
Comment (test-fullautomation) [inline RobotFramework_DoIP/DoipKeywords.py]: Hi Thong,

why did you  not implement this?
Doku says:
reconnect(close_delay=2)
Attempts to re-establish the connection. Useful after an ECU reset

Parameters:
close_delay (float, optional) – Time to wait between closing and re-opening socket

So, I expect the same interface as keyword.

Thank you,
 Thomas

**Review by test-fullautomation — CHANGES_REQUESTED** [review]
Hi Thong,
Thank you for your initial pull-request.

Please refer to my review findings.

Please also implement keywords for all methods. E.g. 
read_doip(timeout=2, transport=TransportType.TRANSPORT_TCP)
is missing.

e.g. it need to be possible by means of keyword to use encrypted communication, ...

* What about DoIP messages and UDS?
* Please also check that setup.py installes the library properly.
* packagedoc need to be done, too. DoIP must be part of our main documentation.
   Base is genpackagedoc. Please discuss with @ngoan1608 how to use.

Thank you,
 Thomas

File: RobotFramework_DoIP/DoipKeywords.py (hunk 1 above)
Comment (huavanthong) [reply RobotFramework_DoIP/DoipKeywords.py]: Hi Thomas, 

I have updated it as the method:
```python
def connect_to_ecu(
        self,
        ecu_ip_address,
        ecu_logical_address,
        tcp_port=TCP_DATA_UNSECURED,
        udp_port=UDP_DISCOVERY,
        activation_type=ActivationTypeDefault,
        protocol_version=0x02,
        client_logical_address=0x0E00,
        client_ip_address=None,
        use_secure=False,
        auto_reconnect_tcp=False,
    ):
```
With this method, we will have many ways to set up a connection for DoIP, then we can interact with ECU by the other methods.

Thank you,
Thong Hua

File: RobotFramework_DoIP/DoipKeywords.py (hunk 3 above)
Comment (huavanthong) [reply RobotFramework_DoIP/DoipKeywords.py]: Hi Thomas,

I have update it as methods:
```python
 @keyword("Reconnect To Ecu")
    def reconnect_to_ecu(self, close_delay=A_PROCESSING_TIME):
```
Please take it for your info.

Thank you,
Thong Hua

File: RobotFramework_DoIP/DoipKeywords.py (hunk 2 above)
Comment (huavanthong) [reply RobotFramework_DoIP/DoipKeywords.py]: Hi Thomas,

I have updated it as methods:
```python
@keyword("Get Entity")
    def get_entity(
        ecu_ip_address="255.255.255.255", 
        protocol_version=0x02, 
        eid=None, 
        vin=None,
    ):
```
If you have any comments, please tell me.

Thank you,
Thong Hua

**Review by test-fullautomation — APPROVED** [review]
Hi Thong,
looks good to me!
Thank you,
 Thomas

**Comment by test-fullautomation** [comment]
Hi  @huavanthong ,
unforunately I have a merge conflict.
Please solve this as soon as possible.
Thank you,
 Thomas

**Comment by huavanthong** [comment]
Hi Thomas,

I have just resolved the conflict. 
Please help me merge it.

Thank you, 
Thong Hua

**Pull request merged**